Cleanup character io interfaces

rmkaplan opened the pull request:
I think this is ready for more general testing and eventual merging.  It isolates the XCCS/NS external format dependencies into one new file XCCS, and all character access then goes through a few abstract functional interfaces (\INCCODE, \PEEKCCODE...), with the XCCS implementations still set up as the default.

This required touching and recompiling lots of files, but it all seems to hang together in this full sysout and in the LFG system built on top.

Other improvements/fixes to MODERNIZE, WHEELSCROLL, CLIPBOARD, and others.

rmkaplan commented:
I have updated to fix some glitches that I observed in further testing.  In particular, the interface to MAKEFILE now lets you specify a new format (like :UTF-8), and that propagates through remaking and also through compiling and recompiling.

Those are new capabilities.  The main strategy for reviewing and testing this branch would be just to run in it.  All old functionality is supposed to work as it used to, despite the changes to low-down interfaces. That's the first confirmation.

It may be appropriate to merge when all old behavior appears to be stable and correct.  Then we can inch up on the new encoding capabilities.

masinter reviewed (commented):
why is sources/LLREAD.LCOM.~1~ on .gitignore? it already ignores *~

masinter reviewed (commented):
there were changes (introductions of functions) to BIGBITMAPS in the library -- not loaded in FULL. Were these not part of the IO functions?

masinter reviewed (commented):
CHAT was recompiled with no changes. CHAT(SHELL) is the only useful CHAT target.
But shouldn't it assume :UTF-8 ?   I think CHAT(host) could turn into ssh but for now it is better to leave out?
CHAT(SHELL) currently always opens /bin/csh (C-shell) which isn't in most Linux distros. See #121

rmkaplan commented:
I think this was something else that got included to make the masterscope analysis more accurate, making sure that all ambiguous field names were qualified.

> On Jul 22, 2021, at 9:12 AM, Larry Masinter ***@***.***> wrote:
> 
> 
> @masinter commented on this pull request.
> 
> there were changes (introductions of functions) to BIGBITMAPS in the library -- not loaded in FULL. Were these not part of the IO functions?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/348#pullrequestreview-712982681>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJORUI5BFR2E7AL7H73TZA7NVANCNFSM47OWLHAA>.
>

rmkaplan commented:
Could be that macros or records changed underneath.

> On Jul 22, 2021, at 9:34 AM, Ron Kaplan ***@***.***> wrote:
> 
> I think this was something else that got included to make the masterscope analysis more accurate, making sure that all ambiguous field names were qualified.
> 
>> On Jul 22, 2021, at 9:12 AM, Larry Masinter ***@***.*** ***@***.***>> wrote:
>> 
>> 
>> @masinter commented on this pull request.
>> 
>> there were changes (introductions of functions) to BIGBITMAPS in the library -- not loaded in FULL. Were these not part of the IO functions?
>> 
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/348#pullrequestreview-712982681>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJORUI5BFR2E7AL7H73TZA7NVANCNFSM47OWLHAA>.
>> 
>

rmkaplan commented:
Dunno

> On Jul 22, 2021, at 8:54 AM, Larry Masinter ***@***.***> wrote:
> 
> 
> @masinter commented on this pull request.
> 
> why is sources/LLREAD.LCOM.1 on .gitignore? it already ignores *~
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/348#pullrequestreview-712963620>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJJHMGXMKTGPEMRJLXLTZA5MPANCNFSM47OWLHAA>.
>

masinter reviewed (commented):
there was a change to SYSEDIT that removed resetting MSMACROPROPS to leave MACRO-FN on the list, leaving the value on MSANALYZE. Macros for all kinds are analyzed by analyzing their expansion. There was also a patch to remove / no longer generate spurious incorrect masterscope "templates" (call to SETTEMPLATE in GETTEMPLATE).
Otherwise,

masinter commented on .gitignore:
```diff
@@ -27,3 +27,4 @@ core
 
 # set up by install-diff-filter.sh script
 .gitattributes
+sources/LLREAD.LCOM.~1~
```
*~ is already gitignored

masinter replied:
now issue #357

masinter commented on library/BIGBITMAPS:
```diff
@@ -1,34 +1,38 @@
 
 
-; Copyright (c) 1991, 1993, 1994 by Venue.  All rights reserved.
+; Copyright (c) 1991, 1993-1994, 2021 by Venue.
```
new copyright could say Interlisp.org

masinter replied:
Issues #207 (copyright dates) and (possibly?) #39 (larger screen bitmap)

masinter commented on library/BIGBITMAPS:
```diff
@@ -1,34 +1,38 @@
                (MACROS |\\SFInvert|))
         (INITRECORDS BIGBM)
-        (FNS BITBLT.BIGBM BITMAPCREATE.BIGBM BITMAPCREATE BITMAPCOPY BLTSHADE.BIGBM BITBLT 
+        (FNS BIGBITMAPP BITBLT.BIGBM BITMAPCREATE.BIGBM BITMAPCREATE BITMAPCOPY BLTSHADE.BIGBM BITBLT
```
is there any documentation for these changes or BIGBITMAP itself?

masinter commented on library/BIGBITMAPS:
```diff
@@ -65,11 +69,11 @@
 
 (PUTPROPS |\\SFInvert| MACRO ((|BitMap| \y)
 
-         (* |corrects| |for| |the| |fact| |that| |alto| |bitmaps| |are| |stored| |with| 
```
can we convert these comments to strings? Or use a different readtable than XCL?

masinter commented on library/BIGBITMAPS:
```diff
@@ -84,6 +88,10 @@
        '6)
 (DEFINEQ
 
+(BIGBITMAPP
```
I suppose these changes are to avoid the cross-file record access. Sorry that just became clear.

rmkaplan commented:
Is this a change that didn’t get captured in this branch?  I don’t remember doing anything about macroprops or templates.

I did fix masterscope to go to the whereis database as a last resort for CONTAINS queries, so I could find the things I needed to work on.

(Also, I extended DUMPDATABASE to take a FILE argument so that it could dump the database with a define-file-info reader environment that would load in all contexts.)

> On Jul 22, 2021, at 12:28 PM, Larry Masinter ***@***.***> wrote:
> 
> 
> @masinter commented on this pull request.
> 
> there was a change to SYSEDIT that removed resetting MSMACROPROPS to leave MACRO-FN on the list, leaving the value on MSANALYZE. Macros for all kinds are analyzed by analyzing their expansion. There was also a patch to remove / no longer generate spurious incorrect masterscope "templates" (call to SETTEMPLATE in GETTEMPLATE).
> Otherwise,
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/348#pullrequestreview-713159483>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJKHYE3WRK3XAWX6BOTTZBWMJANCNFSM47OWLHAA>.
>

rmkaplan commented:
Yes, to export a test that didn’t require the declaration to be present.  (The logic in that code is still a little confusing, I think the bigbitmap extension is wired in at a slightly wrong level.  But I left it.)

> On Jul 22, 2021, at 1:11 PM, Larry Masinter ***@***.***> wrote:
> 
> 
> @masinter commented on this pull request.
> 
> In library/BIGBITMAPS <https://github.com/Interlisp/medley/pull/348#discussion_r675129760>:
> 
> > @@ -84,6 +88,10 @@
>         '6)
>  (DEFINEQ
>  
> +(BIGBITMAPP
> I suppose these changes are to avoid the cross-file record access. Sorry that just became clear.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/348#pullrequestreview-713193333>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJIA23UR437NKX3CJCLTZB3PTANCNFSM47OWLHAA>.
>

masinter commented on lispusers/MODERNIZE:
```diff
@@ -1,11 +1,11 @@
-(DEFINE-FILE-INFO READTABLE "INTERLISP" PACKAGE "INTERLISP")
-(FILECREATED "14-Mar-2021 20:33:34" 
-{DSK}<Users>kaplan>Local>medley3.5>git-medley>lispusers>MODERNIZE.;14 20950  
```
shouldn't this move to Library? Basically anything that is part of FULL.

masinter reviewed (changes requested):
The problem with MSMACROPROPS is just an indication that you made your branch and didn't update it from master for months; I wasn't doing a _lot_ of editing but we need to resolve the conflicts (not many).

I've gone through about 50 / 150 of the changed files in about 5 hours. I'll try to go through the rest.

rmkaplan commented:
Thanks for slogging through this, I know it’s a lot.

In terms of legacy dates, I thought it was set up to do that.  It certainly does that in my LFG environment.  Maybe there is a flag some where that should be set.

In fact, it seems that in LFG I have a file EDITDATEPATCH that has a variable SINGLEEDITDATE that controls that.  Maybe the system edit date isn’t set up that way.

> On Jul 22, 2021, at 8:50 PM, Larry Masinter ***@***.***> wrote:
> 
> 
> @masinter requested changes on this pull request.
> 
> The problem with MSMACROPROPS is just an indication that you made your branch and didn't update it from master for months; I wasn't doing a lot of editing but we need to resolve the conflicts (not many).
> 
> I've gone through about 50 / 150 of the changed files in about 5 hours. I'll try to go through the rest.
> 
> In library/BIGBITMAPS <https://github.com/Interlisp/medley/pull/348#discussion_r675131121>:
> 
> > @@ -1400,14 +1516,14 @@
>            (RETURN COLORBITMAP))))
>  
>  (\\BWTOCOLORBLT
> -  (LAMBDA (SOURCEBWBM SLEFT SBOTTOM DESTCOLORBM DLEFT DBOTTOM WIDTH HEIGHT 0COLOR 1COLOR DESTNBITS)
> -                                                          (* \; "Edited 13-Jul-90 14:11 by matsuda")
> -
> -         (* |blits| |from| \a |black| |and| |white| |bitmap| |into| \a |color| |bitmap| 
> -       |which| |has| DESTNBITS |bits| |per| |pixel.|
> -       DESTCOLORBM |is| \a |pointer| |to| |the| |color| |bitmap.|)
> -                                                             (* |assumes| |all| |datatypes| |and| 
> -                                                           |bounds| |have| |been| |checked|)
> +  (LAMBDA (SOURCEBWBM SLEFT SBOTTOM DESTCOLORBM DLEFT DBOTTOM WIDTH HEIGHT 0COLOR 1COLOR DESTNBITS)
> +                                                           (* \; "Edited  8-May-2021 22:31 by rmk:")
> shouldn't we leave in the "legacy" edit dates?
> 
> In library/BIGBITMAPS <https://github.com/Interlisp/medley/pull/348#discussion_r675132891>:
> 
> > @@ -1454,42 +1570,42 @@
>                                                             DESWRD))
>                                   WIDTH MAP 0COLOR 1COLOR))
>                   (COND
> is there a reason the font-change character changed in a way that makes these changes show up in diffs
> 
> In library/BIGBITMAPS <https://github.com/Interlisp/medley/pull/348#discussion_r675135577>:
> 
> >                             (BITBLT DESTCOLORBM DESALIGNLEFT DBOTTOM DESTCOLORBM (IPLUS 
>                                                                                           DESALIGNLEFT
>                                                                                             DESOFF)
>                                    DBOTTOM WIDTH HEIGHT 'INPUT 'REPLACE)
>                             (BITBLT SCR 0 0 DESTCOLORBM DESALIGNLEFT DBOTTOM DESOFF HEIGHT
>                                    'INPUT
>                                    'REPLACE)))))
> -        (8 ((OPCODES SUBRCALL 142 11)
> it looks like the (OPCODES SUBRCALL ...) was moved? We should look soon at BIGBITMAP/LLCOLOR and document what we find.
> 
> In library/CLIPBOARD <https://github.com/Interlisp/medley/pull/348#discussion_r675143131>:
> 
> > @@ -203,8 +203,8 @@
>  )
>  (PUTPROPS CLIPBOARD COPYRIGHT (NONE))
> copyright Interlisp.org?
> Paste on windows/linux into lisp seems fine but I don't see meta-C copy work.
> the documentation -- such as it is -- says UNIX-GETENV("ostype") but I believe it is case sensitive UNIX-GETENV("OSTYPE") -- which is NIL on my system. This probably needs documentation in the Medley new release notes.
> 
> In library/FILEBROWSER <https://github.com/Interlisp/medley/pull/348#discussion_r675144570>:
> 
> > @@ -1,12 +1,11 @@
> -(DEFINE-FILE-INFO READTABLE "XCL" PACKAGE "INTERLISP")
> -(FILECREATED "27-Feb-2021 20:08:26" 
> -|{DSK}<Users>kaplan>Local>medley3.5>git-medley>library>FILEBROWSER.;33| 261320 
> +(DEFINE-FILE-INFO PACKAGE "INTERLISP" READTABLE "XCL" BASE 10 FORMAT XCCS)
> the bigger DEFINE-FILE-INFO ... is that necessary?
> :BASE 10 ? I'm still uncomfortable with :FORMAT, and with what seems like excessive generality (it's either XCCS-runcode or UTF-8?
> 
> In library/MASTERSCOPE <https://github.com/Interlisp/medley/pull/348#discussion_r675245982>:
> 
> > @@ -2649,10 +2668,10 @@ Copyright (c) 1983-1988, 1990, 1993-1994, 2018, 2020-2021 by Venue & Xerox Corpo
>  (DEFINEQ
>  
>  (MSINTERPRETSET
> -  [LAMBDA (SET OP ARG)                                   (* ; "Edited 15-Aug-90 11:50 by jds")
> All files (just noticed here) legacy editdates should be retained
> 
> In library/MASTERSCOPE <https://github.com/Interlisp/medley/pull/348#discussion_r675246671>:
> 
> > @@ -2759,16 +2778,17 @@ Copyright (c) 1983-1988, 1990, 1993-1994, 2018, 2020-2021 by Venue & Xerox Corpo
>                                  (SHOULDNT 12)))
>                     (QUOTE (SELECTQ OP
>                                (CHECK (COND
> -                                        (ARG (SETQ NEEDUPDATE (UNION NEEDUPDATE (fetch QUOTED
> -                                                                                   of REP)))
> +                                        (ARG (SETQ NEEDUPDATE (UNION NEEDUPDATE (fetch
> +                                                                                 'QUOTED of
> the ' in (fetch 'QUOTED of REP) doesn't look right -- rename the record QUOTE to Q?
> 
> In library/POSTSCRIPTSTREAM <https://github.com/Interlisp/medley/pull/348#discussion_r675249963>:
> 
> >  
>  
>  (* ; "
> -Copyright (c) 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1997, 1998, 2018 by Venue  This program or documentation contains confidential information and trade secrets of Venue.  Reverse engineering, reverse compiling and disassembling of object code are prohibited.  Use of this program or documentation is governed by written agreement with Venue.  Use of copyright notice is precautionary and does not imply publication or disclosure of trade secrets.  All rights reserved.
> +Copyright (c) 1989-1995, 1997-1998, 2018, 2021 by Venue  This program or documentation contains confidential information and trade secrets of Venue.  Reverse engineering, reverse compiling and disassembling of object code are prohibited.  Use of this program or documentation is governed by written agreement with Venue.  Use of copyright notice is precautionary and does not imply publication or disclosure of trade secrets.
> copyright 1989-1995 Venue 2018-2021 Interlisp.org
> 
> Remove "license" statement entirely
> 
> In library/POSTSCRIPTSTREAM <https://github.com/Interlisp/medley/pull/348#discussion_r675252102>:
> 
> >                  "%%%%Title: "
>                  (MKSTRING (OR (LISTGET OPTIONS 'DOCUMENT.NAME)
>                                FILE))
>                  T "%%%%Creator: PostScript Driver Copyright (C) 1988-1992 Venue and others" T 
>                  "%%%%CreationDate: " (DATE)
>                  T %# (COND
> -                        ((EQ 'LPT (FILENAMEFIELD FP 'HOST))
> +                        ((EQ 'LPT (FILENAMEFIELD STREAM 'HOST))
> The break page mechanism is archaic. Creating PDF with Unicode would be better. Does {LPT} work?
> 
> In library/POSTSCRIPTSTREAM <https://github.com/Interlisp/medley/pull/348#discussion_r675254516>:
> 
> > +        ("356,342" SYMBOL "2,375")
> +                                                             (* ; "centerrightbracce")
> +        ("356,343" SYMBOL "2,376")
> +                                                             (* ; "bottomrightbracce")
> +        ("356,344" SYMBOL "2,354")
> +                                                             (* ; "topleftbracce")
> +        ("356,345" SYMBOL "2,356")
> +                                                             (* ; "bottomleftbracce")
> +        ("356,346" SYMBOL "2,355")
> +                                                             (* ; "centerleftbracce")
> +        ("356,355" SYMBOL "2,363")
> +                                                             (* ; "integraltop")
> +        ("356,356" SYMBOL "2,365")
> +                                                             (* ; "integralbottom")
> +        ("356,357" SYMBOL "2,364")
> +                                                             (* ; "integralcenter")))
> Any ideas what's wrong with the prettyprinter ?
> 
> In library/RDSYS <https://github.com/Interlisp/medley/pull/348#discussion_r675256128>:
> 
> > @@ -1,78 +1,78 @@
> -(DEFINE-FILE-INFO READTABLE "INTERLISP" PACKAGE "INTERLISP")
> -(FILECREATED "29-Apr-2021 09:48:22" {DSK}<Users>kaplan>Local>medley3.5>git-medley>tmp>RDSYS.;19
> +(DEFINE-FILE-INFO PACKAGE "INTERLISP" READTABLE "XCL" BASE 10 FORMAT XCCS)
> it doesn't seem right for RDSYS's READTABLE to change from "INTERLISP" to "XCL". Why?
> 
> In library/TEDITHCPY <https://github.com/Interlisp/medley/pull/348#discussion_r675258071>:
> 
> > +                                                             (* ; 
> +                                                           "0.75 inches from bottom, 1 from top")
> +              )
> +        [COMS 
> +              (* ;; "Support for the window-menu's HARDCOPY button, LISTFILES, etc.")
> +
> +              (FNS TEDIT.HARDCOPYFN \TEDIT.HARDCOPY \TEDIT.PRESS.HARDCOPY)
> +              (P (LISTPUT (ASSOC 'CONVERSION (ASSOC 'INTERPRESS PRINTFILETYPES))
> +                        'TEDIT
> +                        (FUNCTION \TEDIT.HARDCOPY)))
> +              (P (LET [(PRESSVALUES (ASSOC 'CONVERSION (ASSOC 'PRESS PRINTFILETYPES]
> +                      (COND (PRESSVALUES                     (* ; 
> +                                                  "Only install PRESS printing if PRESS is loaded.")
> +                                   (LISTPUT PRESSVALUES 'TEDIT (FUNCTION \TEDIT.PRESS.HARDCOPY]
> +        (COMS 
> +              (* ;; "vars for Japanese Line Break")
> an example of code assuming XCCS
> 
> In library/UNICODE <https://github.com/Interlisp/medley/pull/348#discussion_r675267397>:
> 
> > @@ -19,7 +19,7 @@
>                (INITVARS (EXTERNALEOL 'LF))
>                (FNS MAKE-UNICODE-FORMATS)
>                (P (MAKE-UNICODE-FORMATS EXTERNALEOL))
> -              (ADDVARS (*DEFAULT-EXTERNAL-FORMATS* (UNIX :UTF8)))
> +              (ADDVARS (*DEFAULT-EXTERNALFORMATS* (UNIX :UTF8)))
> shouldn't this be UTF-8 and not UTF8?
> why "UNIX"?
> 
> In library/UNICODE <https://github.com/Interlisp/medley/pull/348#discussion_r675268259>:
> 
> > @@ -58,23 +58,18 @@
>                (GLOBALVARS *XCCSTOUNICODE* *UNICODETOXCCS*))
>          (FNS HEXSTRING UTF8HEXSTRING NUTF8CODEBYTES NUTF8STRINGBYTES XTOUSTRING XCCSSTRING)
>          (FNS SHOWCHARS)
> -        (DECLARE%: EVAL@COMPILE DONTCOPY 
> -
> -               (* ;; "This is taken from FILEIO")
> -
> -               (CONSTANTS (LF.EOLC 1)
> -                      (CR.EOLC 0)
> -                      (CRLF.EOLC 2))
> +        (DECLARE%: EVAL@COMPILE DONTCOPY (P (OR (GETP 'EXPORTS.ALL 'FILE)
> compiling is so fast no one will watch. Files that need other files to compile should just load them.
> 
> In library/UNICODE <https://github.com/Interlisp/medley/pull/348#discussion_r675269489>:
> 
> > @@ -132,201 +127,11 @@
>                                                       (LOADBYTE C 0 6)))
>                          ELSE (ERROR "CHARCODE too big for UTF8" C])
>  
> -(UTF8.INCCODEFN
> I'm not sure what's responsible, but loadups take twice as long. This looks not very performant
> 
> In library/UNICODE <https://github.com/Interlisp/medley/pull/348#discussion_r675270528>:
> 
> > -              (SETQ CODE (UNICODE.TRANSLATE CODE *UNICODETOXCCS*)))
> -          (RETURN (IF COUNTP
> -                      THEN (CL:VALUES CODE 0)
> -                    ELSE CODE])
> -
> -(\UTF8.BACKCHARFN
> -  [LAMBDA (STREAM COUNTP)                               (* ; "Edited 26-Jul-2020 10:41 by rmk:")
> -
> -    (* ;; "\BACKFILEPTR is NIL at beginning of FILE, just return COUNT")
> -
> -    (BIND (C _ 0) WHILE (CL:WHEN (\BACKFILEPTR STREAM)
> -                                    (ADD C 1)
> -                                    (EQ 2 (LRSH (\PEEKBIN STREAM)
> -                                                6))) REPEATUNTIL (EQ C 4)
> -       FINALLY (RETURN (AND COUNTP C])
> +(UTF8.INCCODEFN
>   [LAMBDA (STREAM BYTECOUNTVAR BYTECOUNTVAL RAW)        (* ; "Edited 15-Jun-2021 13:35 by rmk:")
>                                                             (* ; "Edited  6-Aug-2020 17:13 by rmk:")
> 
>     (* ;; "Do not do UNICODE to XCSS translation if RAW.")
> 
>     (* ;; "Test for smallp because the stream's End-of-file operation may suppress the error")
> 
>     (LET (BYTE1 BYTE2 BYTE3 BYTE4 CODE (COUNT 1))
>          (SETQ BYTE1 (\BIN STREAM))
> 
>          (* ;; "Distinguish on header bytes (modulo peculiar EOF behavior--the caller will get whatever ended up in BYTE1")
> 
>          (CL:WHEN (SMALLP BYTE1)
>              [SETQ CODE
>               (IF (ILESSP BYTE1 128)
>                   THEN 
> 
>                         (* ;; 
>                       "Test first:  Ascii is the common case.  EOL requires its own translation")
> 
>                         (SELCHARQ BYTE1
>                              (CR (SELECTC (FETCH (STREAM EOLCONVENTION) OF STREAM)
>                                      (CR.EOLC                (* ; "Also eq BYTE1")
>                                               (CHARCODE EOL))
>                                      (CRLF.EOLC (IF (EQ (CHARCODE LF)
>                                                             (\PEEKBIN STREAM T))
>                                                     THEN (\BIN STREAM)
>                                                           (CL:WHEN BYTECOUNTVAR
>                                                               (SETQ BYTECOUNTVAR (IDIFFERENCE 
>                                                                                         BYTECOUNTVAL
>                                                                                         2)))
>                                                           (CHARCODE EOL)
>                                                   ELSE BYTE1))
>                                      BYTE1))
>                              (LF (CL:IF (EQ LF.EOLC (FETCH (STREAM EOLCONVENTION) OF STREAM))
>                                      (CHARCODE EOL)
>                                      BYTE1))
>                              BYTE1)
>                 ELSEIF (IGEQ BYTE1 (LLSH 15 4))
>                   THEN                                   (* ; "4 bytes")
>                         (SETQ BYTE2 (\BIN STREAM))
>                         (CL:WHEN (OR (NOT (SMALLP BYTE2))
>                                      (ILESSP BYTE2 128))
>                             (ERROR "INVALID UTF8 SEQUENCE" (LIST BYTE1 BYTE2)))
>                         (SETQ BYTE3 (\BIN STREAM))
>                         (CL:WHEN (OR (NOT (SMALLP BYTE3))
>                                      (ILESSP BYTE3 128))
>                             (ERROR "INVALID UTF8 SEQUENCE" (LIST BYTE1 BYTE2 BYTE3)))
>                         (SETQ BYTE4 (\BIN STREAM))
>                         (CL:WHEN (OR (NOT (SMALLP BYTE4))
>                                      (ILESSP BYTE4 128))
>                             (ERROR "INVALID UTF8 SEQUENCE" (LIST BYTE1 BYTE2 BYTE3 BYTE4)))
>                         (SETQ COUNT 4)
>                         (LOGOR (LLSH (LOADBYTE BYTE1 0 3)
>                                      18)
>                                (LLSH (LOADBYTE BYTE2 0 6)
>                                      12)
>                                (LLSH (LOADBYTE BYTE3 0 6)
>                                      6)
>                                (LOADBYTE BYTE4 0 6))
>                 ELSEIF (IGEQ BYTE1 (LLSH 7 5))
>                   THEN                                   (* ; "3 bytes")
>                         (SETQ BYTE2 (\BIN STREAM))
>                         (CL:WHEN (OR (NOT (SMALLP BYTE2))
>                                      (ILESSP BYTE2 128))
>                             (ERROR "INVALID UTF8 SEQUENCE" (LIST BYTE1 BYTE2)))
>                         (SETQ BYTE3 (\BIN STREAM))
>                         (CL:WHEN (OR (NOT (SMALLP BYTE3))
>                                      (ILESSP BYTE3 128))
>                             (ERROR "INVALID UTF8 SEQUENCE" (LIST BYTE1 BYTE2 BYTE3)))
>                         (SETQ COUNT 3)
>                         (LOGOR (LLSH (LOADBYTE BYTE1 0 4)
>                                      12)
>                                (LLSH (LOADBYTE BYTE2 0 6)
>                                      6)
>                                (LOADBYTE BYTE3 0 6))
>                 ELSE                                     (* ; "Must be 2 bytes")
>                       (SETQ COUNT 2)
>                       (SETQ BYTE2 (\BIN STREAM))
>                       (CL:WHEN (OR (NOT (SMALLP BYTE2))
>                                    (ILESSP BYTE2 128))
>                           (ERROR "INVALID UTF8 SEQUENCE" (LIST BYTE1 BYTE2)))
>                       (LOGOR (LLSH (LOADBYTE BYTE1 0 5)
>                                    6)
>                              (LOADBYTE BYTE2 0 6])
>          (CL:UNLESS (OR RAW (NOT (SMALLP CODE)))
>              (SETQ CODE (UNICODE.TRANSLATE CODE *UNICODETOXCCS*)))
>          (CL:WHEN BYTECOUNTVAR
>              (SET BYTECOUNTVAR (IDIFFERENCE BYTECOUNTVAL COUNT)))
>          CODE])
> not pretty-printed can't be reviewed in github
> 
> In library/UNICODE.TXT <https://github.com/Interlisp/medley/pull/348#discussion_r675271540>:
> 
> > @@ -36,9 +36,9 @@ The function STREAMPROP obtains or changes the external format of an open stream
>  	
>  In the latter case, the stream's format is changed to :UTF8 and the previous value is returned, in this example it is Medley's historical default format :XCCS.
> UTF-8
> 
> In library/XCCS <https://github.com/Interlisp/medley/pull/348#discussion_r675272314>:
> 
> > @@ -1,412 +0,0 @@
> -(DEFINE-FILE-INFO READTABLE "INTERLISP" PACKAGE "INTERLISP")
> -(FILECREATED " 4-Aug-2020 22:36:36" {DSK}<Users>kaplan>Local>medley3.5>lispcore>library>XCCS.;4 23636  
> XCCS was deleted but XCCS.LCOM is still there?
> 
> In library/lafite/LAFITE <https://github.com/Interlisp/medley/pull/348#discussion_r675273222>:
> 
> > @@ -1,16 +1,21 @@
>  (DEFINE-FILE-INFO PACKAGE "INTERLISP" READTABLE "INTERLISP" BASE 10)
> -(FILECREATED "22-Aug-94 13:00:22" {DSK}<king>export>lispcore>lafite>parc-94>LAFITE.;2 73704  
> not sure about Lafite in Medley being available for open source
> 
> In library/lafite/LAFITE <https://github.com/Interlisp/medley/pull/348#discussion_r675273475>:
> 
> >  
>  
>  (* ; "
> -Copyright (c) 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1993, 1994 by Xerox Corporation and Bolt Beranek and Newman Inc..  All rights reserved.
> +Copyright (c) 1982-1989, 1993-1994, 2021 by Xerox Corporation and Bolt Beranek and Newman Inc..
> might need to move it to another repo
> 
> In lispusers/ISO8859IO <https://github.com/Interlisp/medley/pull/348#discussion_r675275687>:
> 
> > -               (* ;; "From FILEIO")
> -
> -               (CONSTANTS (\NORUNCODE 255))
> -               
> -               (* ;; "From LLCHAR")
> -
> -               (CONSTANTS (NSCHARSETSHIFT 255))
> -               
> -               (* ;; "From LLREAD")
> -
> -               (MACROS \XCCSIN \XCCSPEEK \BACKXCCSCHAR)
> -               
> -               (* ;; "From MODARITH")
> -
> -               (MACROS UNFOLD))))
> +        (DECLARE%: EVAL@COMPILE DONTCOPY (P (OR (GETP 'EXPORTS.ALL 'FILE)
> ***@***.*** (files (source) sysedit) ?
> 
> Are these all useful external formats?
> 
> In lispusers/MODERNIZE <https://github.com/Interlisp/medley/pull/348#discussion_r675278260>:
> 
> >                
>                (* ;; "Add some Meta commands")
>  
>                (FNS TEDIT.MODERNIZE TEDIT.SELECTALL)
> -              (DECLARE%: DONTEVAL@LOAD DOCOPY (P (TEDIT.MODERNIZE)
> +              (DECLARE%: DONTEVAL@LOAD DOCOPY (P 
> Does this work in a LISP.SYSOUT without TEdit loaded?
> 
> In lispusers/MODERNIZE <https://github.com/Interlisp/medley/pull/348#discussion_r675279028>:
> 
> >  
> -    (CL:UNLESS (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN)
> -        (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN (WINDOWPROP WINDOW 'BUTTONEVENTFN))
> -        (WINDOWPROP WINDOW 'BUTTONEVENTFN (IF ANYWHERE
> -                                              THEN (FUNCTION MODERNWINDOW.BUTTONEVENTFN.ANYWHERE)
> -                                            ELSE (FUNCTION MODERNWINDOW.BUTTONEVENTFN))))
> +    (CL:WHEN (AND TITLEPROPORTION (GREATERP TITLEPROPORTION 0.5))
> +           (ERROR "TITLEPROPORTION cannot be greater than .5"))
> +    (CL:WHEN (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN)
> +        (WINDOWPROP WINDOW 'BUTTONEVENTFN (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN))
> +        (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN NIL))
> +    (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN (WINDOWPROP WINDOW 'BUTTONEVENTFN))
> +    (WINDOWPROP WINDOW 'BUTTONEVENTFN (IF (OR ANYWHERE TITLEPROPORTION)
> is some kind of visual feedback possible of which window titlebars and corners can or can't be buttoned?
> Does MODERNIZE work with Rooms?
> 
> In lispusers/MODERNIZE <https://github.com/Interlisp/medley/pull/348#discussion_r675286708>:
> 
> > @@ -1,11 +1,11 @@
> -(DEFINE-FILE-INFO READTABLE "INTERLISP" PACKAGE "INTERLISP")
> -(FILECREATED "14-Mar-2021 20:33:34" 
> -{DSK}<Users>kaplan>Local>medley3.5>git-medley>lispusers>MODERNIZE.;14 20950  
> shouldn't this move to Library? Basically anything that is part of FULL.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/348#pullrequestreview-713195162>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJLCFAJTOOXJOWGQWLTTZDRHJANCNFSM47OWLHAA>.
>

masinter commented on sources/CMLREAD:
```diff
@@ -237,38 +240,41 @@ Copyright (c) 1985, 1986, 1987, 1988, 1990, 1993 by Venue & Xerox Corporation.
                   (*READ-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E))
-                  (*PRINT-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E)))
+                  (*PRINT-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E))
+                  (*EXTERNALFORMAT* (ffetch (READER-ENVIRONMENT REFORMAT) of E)))
```
where are the other variables declared SPECIAL? Why just the external format? Is this really at the right level? The other elements don't depend / vary when you change XCCS and UTF-8... they're EXTERNAL-FORMAT independent

masinter replied:
Issue #371

masinter commented on sources/CMLREAD:
```diff
@@ -237,38 +240,41 @@ Copyright (c) 1985, 1986, 1987, 1988, 1990, 1993 by Venue & Xerox Corporation.
 
 (RPAQ? *COMMON-LISP-READ-ENVIRONMENT* (create READER-ENVIRONMENT REPACKAGE _ (CL:FIND-PACKAGE
                                                                                   "USER")
-                                                 REREADTABLE _ CMLRDTBL REBASE _ 10))
```
same ... common lisp vs. interlisp readtable shouldn't change the "stream" properties.

masinter commented on sources/EXTERNALFORMAT.TXT:
```diff
@@ -0,0 +1,70 @@
+New architecture for character input-output and alternative external formats
```
we really need to put together a Medley 4 release manual, formatted and fixed up

masinter commented on sources/FASLOAD:
```diff
@@ -1,14 +1,14 @@
+(IL:FILECREATED "10-Jun-2021 18:26:43" 
+IL:|{DSK}<Users>kaplan>Local>medley3.5>git-medley>sources>FASLOAD.;8| 35584  
 
-      IL:|changes| IL:|to:|  (IL:FNS CONVERT-FASL-DATE)
```
This is really illegible IL:|changes| IL:|to:|

nbriggs replied:
Happens any time the FILECREATED expressions are written/read with a readtable other than an IL based one.  IL:|previous| IL:|date|, etc.

masinter replied:
a separate issue, but I think we should change the FILECREATED format to use strings or keywords

   "changes to" IL:READ "previous" "12-DEC-2021"

or

   :changes "IL:READ" :previous-date "12-Dec-2021"

so it isn't so ugly. This is in support of "Modernization" and "better support of Common Lisp".

masinter reviewed (changes requested):
while there are a lot of "cosmetic" issues (bad pretty printing, edit comments lost, etc) there are a few more serious issues.  I think it's necessary to sort out the "Masterscope gives error on ambiguous record field" changes from the "support external character encoding". The former could be fixed by adding a per-file record priority list, removing the necessity of any source edits. And the READTABLE should be independent of the file encoding. READ works on a stream of characters, which remain the same character independent of whether they are encoded in XCCS, UTF-8, ISO8059-10 or what have you.
I propose using this branch as a model, and bringing forward changes one at a time starting with current master.

rmkaplan commented:
With respect to CMLREAD, the other variables are the variables that define the current default readtabe, package, and base, when you call IO functions with those not locally specified.  I presume that there are global declarations somewhere that they are specvars.

I added the external format as another variable, when I thought that it was of the same sort and that was the right thing to do.  But it isn’t, the external format is a property of the stream itself, not of the printing environment.  So that variable is really not interpreted—the external format is installed specifically when the stream is opened or the user changes it by STREAMPROP or SETFILEINFO.

To make it operate implicitly as part of the context set by this macro, a further change would have been to have this reset the format of the stream under a resetsave and restore the previous external format on exit.  But the problem is that this macro doesn’t actually take the stream as an argument, it just sets these variables for any and all printing/reading to any and all streams underneath.

(Note also that setting the format changes all 5 of the primitive character reading/writing functions in the stream. It’s a level below whether or not a given internal character is a sepr in this context, or a package identifier should be printed.)

(Note also:  I have no idea why this was in CMLREAD)

> On Jul 23, 2021, at 11:40 AM, Larry Masinter ***@***.***> wrote:
> 
> 
> @masinter requested changes on this pull request.
> 
> while there are a lot of "cosmetic" issues (bad pretty printing, edit comments lost, etc) there are a few more serious issues. I think it's necessary to sort out the "Masterscope gives error on ambiguous record field" changes from the "support external character encoding". The former could be fixed by adding a per-file record priority list, removing the necessity of any source edits. And the READTABLE should be independent of the file encoding. READ works on a stream of characters, which remain the same character independent of whether they are encoded in XCCS, UTF-8, ISO8059-10 or what have you.
> I propose using this branch as a model, and bringing forward changes one at a time starting with current master.
> 
> In lispusers/PRETTYFILEINDEX <https://github.com/Interlisp/medley/pull/348#discussion_r675677958>:
> 
> > @@ -181,7 +183,8 @@ Copyright (c) 1988, 1992, 1993, 1999 by Xerox Corporation.  All rights reserved.
>  (DEFINEQ
>  
>  (PRETTYFILEINDEX
> -  [LAMBDA (FILENAME PRINTOPTIONS OUTSTREAM DONTINDEX)        (* ; "Edited 11-Apr-95 00:02 by rmk:")
> we should fix ALL the copyright notices
> 
> In lispusers/PRETTYFILEINDEX <https://github.com/Interlisp/medley/pull/348#discussion_r675678489>:
> 
> > @@ -181,7 +183,8 @@ Copyright (c) 1988, 1992, 1993, 1999 by Xerox Corporation.  All rights reserved.
>  (DEFINEQ
>  
>  (PRETTYFILEINDEX
> -  [LAMBDA (FILENAME PRINTOPTIONS OUTSTREAM DONTINDEX)        (* ; "Edited 11-Apr-95 00:02 by rmk:")
> +  [LAMBDA (FILENAME PRINTOPTIONS OUTSTREAM DONTINDEX)   (* ; "Edited  9-Jul-2021 21:35 by rmk:")
> ... and edit dates, before they get lost
> 
> In lispusers/WHEELSCROLL <https://github.com/Interlisp/medley/pull/348#discussion_r675685531>:
> 
> >          (GLOBALVARS WHEELSCROLLDELTA WHEELSCROLLSETTLETIME \WHEELSCROLLINPROGRESS)
> -        (INITVARS (WHEELSCROLLDELTA 20)
> +        
> +        (* ;; "To restore the keyactions for PAD..., if the keyaction table has been reinitialized")
> +
> +        [ADDVARS (AFTERSYSOUTFORMS (AND WHEELSCROLLENABLED (ENABLEWHEELSCROLL T)))
> It's not just "after sysout": there are four events: sysout, makesys, logout, savevm which have before, after, and (for all except LOGOUT) "after doing". And this triggers a "DEVICEEVENT" (need to check on the details)
> 
> In lispusers/WHEELSCROLL <https://github.com/Interlisp/medley/pull/348#discussion_r675687734>:
> 
> > -             T)
> -        (521 (WHEELSCROLL 'VERTICAL (IMINUS WHEELSCROLLDELTA))
> -             T)
> -        (522 (WHEELSCROLL 'HORIZONTAL (IMINUS WHEELSCROLLDELTA)
> -                    T))
> -        (523 (WHEELSCROLL 'HORIZONTAL WHEELSCROLLDELTA T))))
> +
> +
> +(* ;; 
> +"These are the highest meta-ctrl characters, they will be unaffected by teh state of ctrl and meta mode keys"
> +)
> +
> +(DECLARE%: EVAL@COMPILE DONTCOPY 
> +(DECLARE%: EVAL@COMPILE 
> +
> +(RPAQQ UP 156)
> i'm wary about using such common words for constants. WS-UP, WS-DOWN etc?
> 
> In sources/ABASIC <https://github.com/Interlisp/medley/pull/348#discussion_r675691283>:
> 
> >                (VARS (\IMAX.FLOAT (FIX MAX.FLOAT))
>                      (\IMIN.FLOAT (FIX MIN.FLOAT)))
>                (GLOBALVARS \IMAX.FLOAT \IMIN.FLOAT))
>          (DECLARE%: DONTEVAL@LOAD DOEVAL@COMPILE DONTCOPY COMPILERVARS
>                 (ADDVARS (NLAMA DUMMYDEF)
>                        (NLAML)
> -                      (LAMA LIST* LIST NCONC NOTIMP ZERO TRUE NILL])
> I wonder if LAMA is needed ? It might be a remnant from Interlisp-10 which compiled calls to (LAMBDA N ...) functions differently than calls to (LAMBDA (...) ...) functions.
> 
> In sources/ABASIC <https://github.com/Interlisp/medley/pull/348#discussion_r675692441>:
> 
> > @@ -62,7 +64,7 @@ Copyright (c) 1982, 1983, 1984, 1985, 1986, 1987, 1990 by Venue & Xerox Corporat
>  (DEFINEQ
>  
>  (NILL
> -  [LAMBDA NOBIND                                             (* lmm " 4-OCT-83 03:05")
> logically, this is the wrong place to fix the NOBIND error. Some bell is ringing about checking common lisp function calls in the ANSI standard Common Lisp
> 
> In sources/ADISPLAY <https://github.com/Interlisp/medley/pull/348#discussion_r675697973>:
> 
> >  
>      (DECLARE (LOCALVARS . T))
> -    [COND
> cant tell what changed here or why...
> 
> In sources/AOFD <https://github.com/Interlisp/medley/pull/348#discussion_r675703991>:
> 
> > @@ -334,9 +335,10 @@ Copyright (c) 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1990 by Venue & Xerox Co
>      (\DEFINEDEVICE NIL \BASEBYTESDEVICE])
>  
>  (\MAKEBASEBYTESTREAM
> -  [LAMBDA (BASE OFFST LEN ACCESS WRITEXTENSIONFN OSTREAM)    (* ; "Edited 17-Jan-87 16:08 by bvm:")
> -          
> -          (* ;; "If an error is to occur due to non-numeric arg or range restrictions, then let it happen outside the UNINTERRUPTABLY")
> +  [LAMBDA (BASE OFFST LEN ACCESS WRITEXTENSIONFN OSTREAM)
> why the pair of font control changes (4 instead of two) father the OSTREM) ? Curious
> 
> In sources/AOFD <https://github.com/Interlisp/medley/pull/348#discussion_r675705469>:
> 
> > -        (replace ACCESS of OSTREAM with ACCESS)
> -          
> -          (* ;; "Insures that the BINABLE BOUTABLE and EXTENDABLE bits are setup setup, and that the correct BIN and BOUT fns are 'inherited' from the FDEV as well")
> -
> -        (freplace FULLFILENAME of OSTREAM with NIL)
> -        (freplace OUTCHARFN of OSTREAM with (FUNCTION \MBS.OUTCHARFN))
> -        (freplace LINELENGTH of OSTREAM with 0)
> -        (freplace CHARPOSITION of OSTREAM with 0)
> -        (freplace WRITEXTENSIONFN of OSTREAM with (SELECTQ ACCESS
> -                                                      ((OUTPUT BOTH) 
> -                                                           WRITEXTENSIONFN)
> -                                                      NIL))
> -        (freplace BBSNCHARS of OSTREAM with 0))
> +        (freplace (STREAM USERCLOSEABLE) of OSTREAM with NIL)
> +        (freplace (STREAM USERVISIBLE) of OSTREAM with NIL)
> +        (freplace (STREAM BYTESIZE) of OSTREAM with BITSPERBYTE)
> BItSPERBYTE is a variable (8 or 9) ?
> 
> In sources/AOFD <https://github.com/Interlisp/medley/pull/348#discussion_r675711019>:
> 
> >  )
>  (DECLARE%: DONTCOPY 
> -(* "FOLLOWING DEFINITIONS EXPORTED")
> -
> -(DECLARE%: EVAL@COMPILE 
> +(* "FOLLOWING DEFINITIONS EXPORTED")(DECLARE%: EVAL@COMPILE 
> odd reformatting of EXPORTS comment
> 
> In sources/ATBL <https://github.com/Interlisp/medley/pull/348#discussion_r675726534>:
> 
> > @@ -1863,24 +1874,24 @@ Copyright (c) 1982, 1983, 1984, 1985, 1986, 1987, 1990, 1993, 2018, 2021 by Venu
>  (PUTPROPS ATBL COPYRIGHT ("Venue & Xerox Corporation" 3676Q 3677Q 3700Q 3701Q 3702Q 3703Q 3706Q 3711Q 
> why are numbers in this file printed in Octal?
> 
> In sources/CMLREAD <https://github.com/Interlisp/medley/pull/348#discussion_r675747061>:
> 
> >  
>  (DEFMACRO WITH-READER-ENVIRONMENT (ENV . BODY)
>     `((CL:LAMBDA (E)
>              (LET ((*PACKAGE* (ffetch (READER-ENVIRONMENT REPACKAGE) of E))
>                    (*READTABLE* (ffetch (READER-ENVIRONMENT REREADTABLE) of E))
>                    (*READ-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E))
> -                  (*PRINT-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E)))
> +                  (*PRINT-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E))
> +                  (*EXTERNALFORMAT* (ffetch (READER-ENVIRONMENT REFORMAT) of E)))
> where are the other variables declared SPECIAL? Why just the external format? Is this really at the right level? The other elements don't depend / vary when you change XCCS and UTF-8... they're EXTERNAL-FORMAT independent
> 
> In sources/CMLREAD <https://github.com/Interlisp/medley/pull/348#discussion_r675749729>:
> 
> >  (DECLARE%: DOEVAL@COMPILE DONTCOPY
>  
>  (GLOBALVARS *COMMON-LISP-READ-ENVIRONMENT*)
>  )
>  
>  (RPAQ? *COMMON-LISP-READ-ENVIRONMENT* (create READER-ENVIRONMENT REPACKAGE _ (CL:FIND-PACKAGE
>                                                                                    "USER")
> -                                                 REREADTABLE _ CMLRDTBL REBASE _ 10))
> same ... common lisp vs. interlisp readtable shouldn't change the "stream" properties.
> 
> In sources/EXTERNALFORMAT.TXT <https://github.com/Interlisp/medley/pull/348#discussion_r675760176>:
> 
> > @@ -0,0 +1,70 @@
> +New architecture for character input-output and alternative external formats
> we really need to put together a Medley 4 release manual, formatted and fixed up
> 
> In sources/FASLOAD <https://github.com/Interlisp/medley/pull/348#discussion_r675763222>:
> 
> >  
> -      IL:|changes| IL:|to:|  (IL:FNS CONVERT-FASL-DATE)
> This is really illegible IL:|changes| IL:|to:|
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/348#pullrequestreview-713878629>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJMCWH24XW4ARKNLIL3TZGZTLANCNFSM47OWLHAA>.
>

rmkaplan commented:
Also, the externalformat.txt document reflects an earlier stage of the migration from what I inherited and where I ended up.  I’ll do an update when this settles down.  (After this merge, there are a few minor cleanups (rationalizing a few of the naming conventions, maybe even taking the format variable out of the WITH-READER-ENVIRONMENT) still to do.)

> On Jul 23, 2021, at 11:40 AM, Larry Masinter ***@***.***> wrote:
> 
> 
> @masinter requested changes on this pull request.
> 
> while there are a lot of "cosmetic" issues (bad pretty printing, edit comments lost, etc) there are a few more serious issues. I think it's necessary to sort out the "Masterscope gives error on ambiguous record field" changes from the "support external character encoding". The former could be fixed by adding a per-file record priority list, removing the necessity of any source edits. And the READTABLE should be independent of the file encoding. READ works on a stream of characters, which remain the same character independent of whether they are encoded in XCCS, UTF-8, ISO8059-10 or what have you.
> I propose using this branch as a model, and bringing forward changes one at a time starting with current master.
> 
> In lispusers/PRETTYFILEINDEX <https://github.com/Interlisp/medley/pull/348#discussion_r675677958>:
> 
> > @@ -181,7 +183,8 @@ Copyright (c) 1988, 1992, 1993, 1999 by Xerox Corporation.  All rights reserved.
>  (DEFINEQ
>  
>  (PRETTYFILEINDEX
> -  [LAMBDA (FILENAME PRINTOPTIONS OUTSTREAM DONTINDEX)        (* ; "Edited 11-Apr-95 00:02 by rmk:")
> we should fix ALL the copyright notices
> 
> In lispusers/PRETTYFILEINDEX <https://github.com/Interlisp/medley/pull/348#discussion_r675678489>:
> 
> > @@ -181,7 +183,8 @@ Copyright (c) 1988, 1992, 1993, 1999 by Xerox Corporation.  All rights reserved.
>  (DEFINEQ
>  
>  (PRETTYFILEINDEX
> -  [LAMBDA (FILENAME PRINTOPTIONS OUTSTREAM DONTINDEX)        (* ; "Edited 11-Apr-95 00:02 by rmk:")
> +  [LAMBDA (FILENAME PRINTOPTIONS OUTSTREAM DONTINDEX)   (* ; "Edited  9-Jul-2021 21:35 by rmk:")
> ... and edit dates, before they get lost
> 
> In lispusers/WHEELSCROLL <https://github.com/Interlisp/medley/pull/348#discussion_r675685531>:
> 
> >          (GLOBALVARS WHEELSCROLLDELTA WHEELSCROLLSETTLETIME \WHEELSCROLLINPROGRESS)
> -        (INITVARS (WHEELSCROLLDELTA 20)
> +        
> +        (* ;; "To restore the keyactions for PAD..., if the keyaction table has been reinitialized")
> +
> +        [ADDVARS (AFTERSYSOUTFORMS (AND WHEELSCROLLENABLED (ENABLEWHEELSCROLL T)))
> It's not just "after sysout": there are four events: sysout, makesys, logout, savevm which have before, after, and (for all except LOGOUT) "after doing". And this triggers a "DEVICEEVENT" (need to check on the details)
> 
> In lispusers/WHEELSCROLL <https://github.com/Interlisp/medley/pull/348#discussion_r675687734>:
> 
> > -             T)
> -        (521 (WHEELSCROLL 'VERTICAL (IMINUS WHEELSCROLLDELTA))
> -             T)
> -        (522 (WHEELSCROLL 'HORIZONTAL (IMINUS WHEELSCROLLDELTA)
> -                    T))
> -        (523 (WHEELSCROLL 'HORIZONTAL WHEELSCROLLDELTA T))))
> +
> +
> +(* ;; 
> +"These are the highest meta-ctrl characters, they will be unaffected by teh state of ctrl and meta mode keys"
> +)
> +
> +(DECLARE%: EVAL@COMPILE DONTCOPY 
> +(DECLARE%: EVAL@COMPILE 
> +
> +(RPAQQ UP 156)
> i'm wary about using such common words for constants. WS-UP, WS-DOWN etc?
> 
> In sources/ABASIC <https://github.com/Interlisp/medley/pull/348#discussion_r675691283>:
> 
> >                (VARS (\IMAX.FLOAT (FIX MAX.FLOAT))
>                      (\IMIN.FLOAT (FIX MIN.FLOAT)))
>                (GLOBALVARS \IMAX.FLOAT \IMIN.FLOAT))
>          (DECLARE%: DONTEVAL@LOAD DOEVAL@COMPILE DONTCOPY COMPILERVARS
>                 (ADDVARS (NLAMA DUMMYDEF)
>                        (NLAML)
> -                      (LAMA LIST* LIST NCONC NOTIMP ZERO TRUE NILL])
> I wonder if LAMA is needed ? It might be a remnant from Interlisp-10 which compiled calls to (LAMBDA N ...) functions differently than calls to (LAMBDA (...) ...) functions.
> 
> In sources/ABASIC <https://github.com/Interlisp/medley/pull/348#discussion_r675692441>:
> 
> > @@ -62,7 +64,7 @@ Copyright (c) 1982, 1983, 1984, 1985, 1986, 1987, 1990 by Venue & Xerox Corporat
>  (DEFINEQ
>  
>  (NILL
> -  [LAMBDA NOBIND                                             (* lmm " 4-OCT-83 03:05")
> logically, this is the wrong place to fix the NOBIND error. Some bell is ringing about checking common lisp function calls in the ANSI standard Common Lisp
> 
> In sources/ADISPLAY <https://github.com/Interlisp/medley/pull/348#discussion_r675697973>:
> 
> >  
>      (DECLARE (LOCALVARS . T))
> -    [COND
> cant tell what changed here or why...
> 
> In sources/AOFD <https://github.com/Interlisp/medley/pull/348#discussion_r675703991>:
> 
> > @@ -334,9 +335,10 @@ Copyright (c) 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1990 by Venue & Xerox Co
>      (\DEFINEDEVICE NIL \BASEBYTESDEVICE])
>  
>  (\MAKEBASEBYTESTREAM
> -  [LAMBDA (BASE OFFST LEN ACCESS WRITEXTENSIONFN OSTREAM)    (* ; "Edited 17-Jan-87 16:08 by bvm:")
> -          
> -          (* ;; "If an error is to occur due to non-numeric arg or range restrictions, then let it happen outside the UNINTERRUPTABLY")
> +  [LAMBDA (BASE OFFST LEN ACCESS WRITEXTENSIONFN OSTREAM)
> why the pair of font control changes (4 instead of two) father the OSTREM) ? Curious
> 
> In sources/AOFD <https://github.com/Interlisp/medley/pull/348#discussion_r675705469>:
> 
> > -        (replace ACCESS of OSTREAM with ACCESS)
> -          
> -          (* ;; "Insures that the BINABLE BOUTABLE and EXTENDABLE bits are setup setup, and that the correct BIN and BOUT fns are 'inherited' from the FDEV as well")
> -
> -        (freplace FULLFILENAME of OSTREAM with NIL)
> -        (freplace OUTCHARFN of OSTREAM with (FUNCTION \MBS.OUTCHARFN))
> -        (freplace LINELENGTH of OSTREAM with 0)
> -        (freplace CHARPOSITION of OSTREAM with 0)
> -        (freplace WRITEXTENSIONFN of OSTREAM with (SELECTQ ACCESS
> -                                                      ((OUTPUT BOTH) 
> -                                                           WRITEXTENSIONFN)
> -                                                      NIL))
> -        (freplace BBSNCHARS of OSTREAM with 0))
> +        (freplace (STREAM USERCLOSEABLE) of OSTREAM with NIL)
> +        (freplace (STREAM USERVISIBLE) of OSTREAM with NIL)
> +        (freplace (STREAM BYTESIZE) of OSTREAM with BITSPERBYTE)
> BItSPERBYTE is a variable (8 or 9) ?
> 
> In sources/AOFD <https://github.com/Interlisp/medley/pull/348#discussion_r675711019>:
> 
> >  )
>  (DECLARE%: DONTCOPY 
> -(* "FOLLOWING DEFINITIONS EXPORTED")
> -
> -(DECLARE%: EVAL@COMPILE 
> +(* "FOLLOWING DEFINITIONS EXPORTED")(DECLARE%: EVAL@COMPILE 
> odd reformatting of EXPORTS comment
> 
> In sources/ATBL <https://github.com/Interlisp/medley/pull/348#discussion_r675726534>:
> 
> > @@ -1863,24 +1874,24 @@ Copyright (c) 1982, 1983, 1984, 1985, 1986, 1987, 1990, 1993, 2018, 2021 by Venu
>  (PUTPROPS ATBL COPYRIGHT ("Venue & Xerox Corporation" 3676Q 3677Q 3700Q 3701Q 3702Q 3703Q 3706Q 3711Q 
> why are numbers in this file printed in Octal?
> 
> In sources/CMLREAD <https://github.com/Interlisp/medley/pull/348#discussion_r675747061>:
> 
> >  
>  (DEFMACRO WITH-READER-ENVIRONMENT (ENV . BODY)
>     `((CL:LAMBDA (E)
>              (LET ((*PACKAGE* (ffetch (READER-ENVIRONMENT REPACKAGE) of E))
>                    (*READTABLE* (ffetch (READER-ENVIRONMENT REREADTABLE) of E))
>                    (*READ-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E))
> -                  (*PRINT-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E)))
> +                  (*PRINT-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E))
> +                  (*EXTERNALFORMAT* (ffetch (READER-ENVIRONMENT REFORMAT) of E)))
> where are the other variables declared SPECIAL? Why just the external format? Is this really at the right level? The other elements don't depend / vary when you change XCCS and UTF-8... they're EXTERNAL-FORMAT independent
> 
> In sources/CMLREAD <https://github.com/Interlisp/medley/pull/348#discussion_r675749729>:
> 
> >  (DECLARE%: DOEVAL@COMPILE DONTCOPY
>  
>  (GLOBALVARS *COMMON-LISP-READ-ENVIRONMENT*)
>  )
>  
>  (RPAQ? *COMMON-LISP-READ-ENVIRONMENT* (create READER-ENVIRONMENT REPACKAGE _ (CL:FIND-PACKAGE
>                                                                                    "USER")
> -                                                 REREADTABLE _ CMLRDTBL REBASE _ 10))
> same ... common lisp vs. interlisp readtable shouldn't change the "stream" properties.
> 
> In sources/EXTERNALFORMAT.TXT <https://github.com/Interlisp/medley/pull/348#discussion_r675760176>:
> 
> > @@ -0,0 +1,70 @@
> +New architecture for character input-output and alternative external formats
> we really need to put together a Medley 4 release manual, formatted and fixed up
> 
> In sources/FASLOAD <https://github.com/Interlisp/medley/pull/348#discussion_r675763222>:
> 
> >  
> -      IL:|changes| IL:|to:|  (IL:FNS CONVERT-FASL-DATE)
> This is really illegible IL:|changes| IL:|to:|
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/348#pullrequestreview-713878629>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJMCWH24XW4ARKNLIL3TZGZTLANCNFSM47OWLHAA>.
>

masinter commented:
Can we at least (a) implement some "unit tests" that makes sure the UTF-8 code is working? Preferably something that can be automated? 
(b) agree to make up "issues" that cover the problems I identified, as a way of getting agreement on what problems you agree need fixing and why? 
I didn't finish reviewing all of the files, and those that diff well (because of pretty printing) are an issue

masinter commented on library/CLIPBOARD:
```diff
@@ -203,8 +203,8 @@
 )
 (PUTPROPS CLIPBOARD COPYRIGHT (NONE))
```
See notes in #211

masinter commented on library/POSTSCRIPTSTREAM:
```diff
@@ -1113,59 +1111,63 @@ Copyright (c) 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1997, 1998, 2018 by Venu
                 "%%%%CreationDate: " (DATE)
                 T %# (COND
-                        ((EQ 'LPT (FILENAMEFIELD FP 'HOST))
+                        ((EQ 'LPT (FILENAMEFIELD STREAM 'HOST))
```
#363 moderrnize printing

masinter commented on lispusers/ISO8859IO:
```diff
@@ -35,23 +35,10 @@ Copyright (c) 1995-1997, 1999, 2021 by Xerox Corporation.
-               (* ;; "From MODARITH")
-
-               (MACROS UNFOLD))))
+        (DECLARE%: EVAL@COMPILE DONTCOPY (P (OR (GETP 'EXPORTS.ALL 'FILE)
```
#365

masinter commented on lispusers/MODERNIZE:
```diff
@@ -92,19 +101,28 @@
+        (WINDOWPROP WINDOW 'BUTTONEVENTFN (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN))
+        (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN NIL))
+    (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN (WINDOWPROP WINDOW 'BUTTONEVENTFN))
+    (WINDOWPROP WINDOW 'BUTTONEVENTFN (IF (OR ANYWHERE TITLEPROPORTION)
```
issue #367

masinter commented:
no longer needed; was turned to draft after merging to make sure issues raised / tested
